File: 205. Isomorphic Strings/205.js
// 205. Isomorphic Strings solution in JavaScript
// https://leetcode.com/problems/isomorphic-strings/description/

/**
 * @param {string} s
 * @param {string} t
 * @return {boolean}
 */
var isIsomorphic = function(s, t) {
    if (s.length != t.length) return false;

    let char_map = new Map();
    let char_set = new Set();

    for (let i = 0; i < s.length; i++) {
        let char_s = s[i];
        let char_t = t[i];

        if (!char_map.get(char_s)) {
            if (char_set.has(char_t)) return false;

            char_map.set(char_s, char_t);
            char_set.add(char_t);
        } else if (char_map.get(char_s)) {
            if (char_map.get(char_s) != char_t) return false;
        }
    }

    return true;
};

File: 205. Isomorphic Strings/205.java
// 205. Isomorphic Strings solution in Java
// https://leetcode.com/problems/isomorphic-strings/description/

class Solution {
    public boolean isIsomorphic(String s, String t) {
        
        int len = s.length();
        HashMap<Character, Character> map_st = new HashMap<>();
        HashMap<Character, Character> map_ts = new HashMap<>();
        
        for (int i = 0; i < len; i++) {
            Character cs = s.charAt(i);
            Character ct = t.charAt(i);
            
            if (map_st.get(cs) != null || map_ts.get(ct) != null) {
                if (map_st.get(cs) != ct) return false;
                if (map_ts.get(ct) != cs) return false;
            } else {
                map_st.put(cs, ct);
                map_ts.put(ct, cs);
            }
        }
        return true;
    }
}